### PR TITLE
Java client

### PR DESCRIPTION
**Pull request by kwf**
This PR merges my work on the java client. It only touches the existing Haskell codebase slightly, changing the error ID numbers for several Cryptol errors, as well as simplifying the implementation of `extractSteps` a tad.

### CODE REVIEW

**Comment by david-christiansen** [comment]
Sorry for the lack of review action here - I'm getting on it right now.

File: java/src/main/java/com/galois/cryptol/client/connection/netstring/Netstring.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
+package com.galois.cryptol.client.connection.netstring;
+
+import java.util.ArrayList;
+import java.util.ArrayDeque;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.io.*;
+
+public class Netstring {
```
[the remark this answers is not in the document]
Comment (david-christiansen) [reply java/src/main/java/com/galois/cryptol/client/connection/netstring/Netstring.java]: This version also looks quite reasonable!

**Comment by david-christiansen** [comment]
I think that the only thing here I'm not totally comfortable with is that the state field became optional. Should we have a chat about this?

File: hello/build.gradle [hunk 2]
```diff
@@ -0,0 +1,19 @@
+apply plugin: 'java'
+apply plugin: 'application'
+
+mainClassName = 'hello.HelloWorld'
```
[the remark this answers is not in the document]
Comment (kwf) [reply hello/build.gradle]: This is a mistaken leftover from a first attempt to set up gradle—I must have missed it during cleanup.

File: hello/src/main/java/hello/.#Greeter.java [hunk 3]
```diff
@@ -0,0 +1 @@
+kwf@000329-kwf.local.23041
```
[the remark this answers is not in the document]
Comment (kwf) [reply hello/src/main/java/hello/.#Greeter.java]: Yup, thanks for catching that. I think this is some kind of Emacs-generated backup file.

File: java/src/main/java/com/galois/cryptol/client/Main.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    // Visual demo of concurrent keyed channel: run a receiving and a sending
+    // thread per channel, each with random delay between send() and request()
+    // calls, displaying the method calls in a table. The simulation lasts for
+    // the timeout parameter, in seconds
+    public static void multiQueueDemo(int channelCount, double meanDelay, int timeout) {
```
[the remark this answers is not in the document]
Comment (kwf) [reply java/src/main/java/com/galois/cryptol/client/Main.java]: This is testing code which I used to debug a tricky set of intermittent race conditions—I think it ought to be moved somewhere.

File: java/src/main/java/com/galois/cryptol/client/CryptolConnection.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        connection.close();
+    }
+
+    // Since this runs on actual output/input streams, we know that connection
+    // exceptions in this case are really IOExceptions, so we use this wrapper
+    // to allow the caller to not need to see ConnectionExceptions
+    private <O> O call(String method, JsonValue params,
+                       Function<JsonValue, O> decode)
+        throws CryptolException {
+        Call<O, CryptolException> call =
```
[the remark this answers is not in the document]
Comment (kwf) [reply java/src/main/java/com/galois/cryptol/client/CryptolConnection.java]: `Call::invoke` perhaps?

File: java/src/main/java/com/galois/cryptol/client/CryptolConnection.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    public CryptolConnection(String server, File dir) throws IOException {
+        this.connection =
+            new Connection(
+                new ProcessBuilder(server, "--dynamic4").directory(dir),
+                (_in, out, err) -> {
+                    // The process will tell us what port to connect to...
+                    int port = (new Scanner(out)).skip("PORT ").nextInt();
+                    // Consume the remaining output and error
+                    forLinesAsync(out, l -> { });
+                    forLinesAsync(err, l -> System.err.println("\u001B[31m" + l + "\u001B[0m"));
```
[the remark this answers is not in the document]
Comment (kwf) [reply java/src/main/java/com/galois/cryptol/client/CryptolConnection.java]: Goodness, it'd be nice for me too! I think they mean "red" :P

File: hello/src/main/java/Netstring.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
+import java.util.ArrayList;
+import java.util.ArrayDeque;
+import java.util.Iterator;
+
+public class Netstring {
```
[the remark this answers is not in the document]
Comment (kwf) [reply hello/src/main/java/Netstring.java]: I think the `hello`-namespaced files are all stale and should be killed. I'll confirm shortly.

File: argo/src/Argo/HistoryWrapper.hs [hunk 8]
```diff
@@ -130,13 +117,13 @@ stateKey = "state"
...
-extractSteps :: Value -> Maybe ([(Text, Value)], Value)
+extractSteps :: Value -> ([(Text, Value)], Value)
 extractSteps v
   | Object o      <- v
   , Just history  <- HashMap.lookup stateKey o
   , Success steps <- fromJSON history
-  , let v' = Object (HashMap.delete stateKey o) = Just (steps, v')
-extractSteps _ = Nothing
+  , let v' = Object (HashMap.delete stateKey o) = (steps, v')
+extractSteps v = ([], v)  -- no state field, so default to empty history
```
[the remark this answers is not in the document]
Comment (kwf) [reply argo/src/Argo/HistoryWrapper.hs]: Of the advantages you cite, (1) was my primary motivation; that is, a desire to have the API really be parametric in the state token.

> I think that not sending a state is just as likely to reflect a misunderstanding of the API. So the potential downside is a prolonged period of confusion before a user realizes that the state is actually mandatory in almost all cases.

Hm, that's probably true.

> Assuming that you're solving the problem of the initial state being basically arbitrary, what if we instead specify that the initial state is the JSON null value, but require that the field be present in the object?

That seems like a good approach.

There was some specific reason that motivated me to make this change on the Haskell side, but I can't recall what it was. I want to make sure I recall it before we converge on another solution.

File: hello/src/main/java/Netstring.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    private final ArrayList<Byte> contents;
```
[the remark this answers is not in the document]
Comment (kwf) [reply hello/src/main/java/Netstring.java]: Per https://github.com/GaloisInc/argo/pull/8#discussion_r322452743, I believe this file is stale—that being said your criticism is valid.

File: java/src/main/java/com/galois/cryptol/client/connection/Call.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
+package com.galois.cryptol.client.connection;
+
+import java.util.function.*;
+
+import com.eclipsesource.json.*;
+import com.galois.cryptol.client.connection.*;
+
+public class Call<O, E extends Exception> extends Notification {
```
[the remark this answers is not in the document]
Comment (kwf) [reply java/src/main/java/com/galois/cryptol/client/connection/Call.java]: Yes, this is a generic remote procedure call, not specific to Cryptol. Which leads me to think: this project really should be split into two libraries: an `argo` interface, and a Cryptol client which uses that interface. What do you think?

File: hello/build.gradle (hunk 2 above)
[the remark this answers is not in the document]
Comment (kwf) [reply hello/build.gradle]: Deleted!

File: hello/src/main/java/Netstring.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (kwf) [reply hello/src/main/java/Netstring.java]: Deleted!

File: hello/src/main/java/Netstring.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (kwf) [reply hello/src/main/java/Netstring.java]: Deleted!

File: hello/src/main/java/hello/.#Greeter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kwf) [reply hello/src/main/java/hello/.#Greeter.java]: Deleted!

**Comment by kwf** [comment]
> Should we have the gradlewrapper.jar in the repository? I wouldn't typically check in binaries, but I don't know enough about the Gradle build process to say for sure here.

I don't know either. I didn't commit it intentionally. Perhaps we should ask someone who knows more about java build tooling than I do—I know sometimes there are weird conventions in situations like this about what you check in.

**Comment by kwf** [comment]
> I don't know very much about Java concurrent programming, so I don't have much useful feedback for you there.

I don't know if it's perfectly idiomatic, but I have tested it pretty reasonably (especially the implementation of the named queue) and it seems race-condition-free even when you hammer on it. Perhaps we should ask someone with more knowledge about idiomatic concurrent java, if we get a chance.

> Overall, there's a lot of Java comments that could just as well be Javadoc, which would make the API more useful to consumers. I don't think that should block a merge, though.

Yes, I think that is a good idea. I will definitely plan to do that.

> I've updated the branch with all the latest changes, but I don't know how to test it. Can you check that I didn't break the Java code?

For future reference, this will run the example application:

```
$ gradle run --console=plain --args='../cryptol-remote-api/.stack-work/dist/x86_64-osx/Cabal-2.4.0.1/build/cryptol-remote-api/cryptol-remote-api ../python'
```

You'll need to replace part of that first path if you are not on macOS, to point to wherever the build artifact gets dropped on your own system.

**Comment by dmzimmerman** [comment]
I haven't looked at it yet, but "it seems race-condition-free even when you hammer on it" jumps out to me, as far as statements about the correctness of concurrent code go. I unfortunately don't have a ton of time to look at it in the immediate term, either.

**Comment by kwf** [comment]
> I haven't looked at it yet, but "it seems race-condition-free even when you hammer on it" jumps out at me, as far as statements about the correctness of concurrent code go. I unfortunately don't have a ton of time to look at it in the immediate term, either.

To clarify: There was a series of race conditions I fixed. I thought about it a lot and wrote some random-testing code that tries to create the conditions for those errors to appear, and I was unable to recreate them after many attempts. I feel pretty confident that at least those errors are not going to reappear.

File: java/src/main/java/com/galois/cryptol/client/Main.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (kwf) [reply java/src/main/java/com/galois/cryptol/client/Main.java]: Moved to `test/` in 6a309ad.

**Comment by kwf** [comment]
On the subject of race conditions: I just re-ran my race condition test code, and it now throws sporadic exceptions when the request queue is closed. In the event of a server crash/restart, I am concerned that these failures will result in some requests being dropped on the floor, which will manifest as infinite hangs on API calls in the event that the server crashes at just the right moment during the call. I don't know if it's a good idea to merge until I sort out this issue, perhaps by reducing the degree of concurrent scalability in favor of simplicity (one global lock that is only held briefly ought to perform well enough in the use cases I am imagining). Let's discuss further next week.

**Comment by david-christiansen** [comment]
Based on a call, @kwf and I decided that this Java code should be placed in a subdirectory with a README describing its current state. The Haskell changes should be dropped, because they're now out of date.

**Pull request closed without merging**